### ISSUE TITLE
Linking containers doesn't work on host machine, but works anywhere else

### ISSUE THREAD

**Issue by CGavrila** [docker/docker#23096]
I am pasting the `docker version` and `docker info` output for both my local machine and a virtual server. 

**Output of `docker version`:**
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**
```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 10
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 99
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 4.2.0-36-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.59 GiB
Name: jmen
ID: XJLM:JFC3:54JV:EO47:CIOK:BKFY:D2MO:NSRS:7EO2:KJCC:PEJZ:ZVDD
WARNING: No swap limit support
```

The above config is for my host machine (my laptop), and the following is for a virtual server in the cloud:

 **Output of `docker version`:**
```
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64
```
**Output of `docker info`:**
```
Containers: 16
 Running: 6
 Paused: 0
 Stopped: 10
Images: 26
Server Version: 1.10.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 241
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: null host bridge
Kernel Version: 3.19.0-59-generic
Operating System: Ubuntu 14.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.86 GiB
Name: backend
ID: TPEU:MDO7:AMPB:PPLD:6JBG:TKW7:T3SX:7QK6:47DH:HFY6:K7XG:KGEJ
WARNING: No swap limit support
```

**Steps to reproduce the issue:**
1.  Use the following `docker-compose.yml` file:
```
version: '2'
services:
  ubuntu:
    image: ubuntu:14.04
    command: "sleep 1000"
    container_name: ubuntu
    links:
      - "redis:redis"
  redis:
    image: redis:3.2.0-alpine
    container_name: redis
    ports:
      - "6379:6379"
```
2. Try to access port 6379 on the `redis` container from the `ubuntu` container. (e.g. `nc -vz redis 6379`)
3. This works fine on my virtual server, but fails on my local machine even though the docker version and config is the same.

**Describe the results you received:**
Can't reach port 6379 on the `redis` container from the `ubuntu` container.

**Describe  the results you expected:**
To be able to access port 6379 in the `redis` container from the `ubuntu` container.

**Comment by HackToday** [comment]
what error info when you access 6379 port ?  
and what's your `iptables -t nat -L -n` and `docker port  <container>` output ?

**Comment by CGavrila** [comment]
```

**Comment by HackToday** [comment]
Could you use  `telnet redis 6379`  to check if it can access 6379 port ?

I think `cat /etc/hosts` would include related redis dns related info, and that link should make it ping that available. Did you check that ?

**Comment by thaJeztah** [comment]
@CGavrila I'm not able to reproduce this. Can you try if you are still able to reproduce on the current version (docker 1.11.1)?

If that doesn't work, can you try simplifying your repro case even more?

First, removing options that are not needed for docker-compose (there's no need to name the container `redis`, as docker-compose automatically adds an alias for the service name. Also, the `link` isn't required, because all containers in the project are part of the same network;

```yaml
version: '2'
services:
  jmen-scraper:
    image: ubuntu:14.04
    command: "sleep 1000"
  redis:
    image: redis:3.2.0-alpine
```

Then, test if you can ping the `redis` service;

```bash
# first do a "docker-compose down" to remove existing containers
docker-compose down
docker-compose exec jmen-scraper sh -c 'ping -c3 redis'

# PING redis (172.21.0.2) 56(84) bytes of data.
# 64 bytes from redis.repro23096_default (172.21.0.2): icmp_seq=1 ttl=64 time=0.073 ms
# 64 bytes from redis.repro23096_default (172.21.0.2): icmp_seq=2 ttl=64 time=0.054 ms
# 64 bytes from redis.repro23096_default (172.21.0.2): icmp_seq=3 ttl=64 time=0.058 ms


docker-compose exec jmen-scraper sh -c 'nc -vz redis 6379'
# Connection to redis 6379 port [tcp/*] succeeded!
```

**Comment by CGavrila** [comment]
@thaJeztah Thanks for taking the time to look into this.

I was probably not explicit enough in the original post, my bad.

So the problem is that I am getting conflicting results on two different hosts. Initially, things looks like this:
```

**Comment by CGavrila** [comment]
@HackToday 
I have tried `telnet` and can't connect to redis:6379. As for `cat /etc/hosts`, there is no explicit `redis` entry, but pinging `redis` works.

@thaJeztah Thanks for taking the time to look into this. I was probably not explicit enough in the original post, my bad.

So the problem is that I am getting conflicting results on two different hosts. Initially, things looks like this:
```
Laptop: docker -v => 1.11.1
VM:       docker -v => 1.10.3
```
so I just assumed that the Docker version must be at fault, so I downgraded the Docker version on my laptop to `1.10.3` and I still get the exact same issue.

So now I have Docker version `1.10.3` on both hosts. On one of them I can access the Redis image the way you'd expect, but on the other machine, pinging redis (`ping -c3 redis`) works, but `nc -vz redis 6379` doesn't. I am using the same `docker-compose.yml` file on both.

So basically, I am looking for some external factor that might influence how the containers talk to each other (e.g. some OS networking setting or similar). I would not expect anyone to be able to reproduce it easily, since this is non-standard behaviour.

**Comment by thaJeztah** [comment]
Ah, right, hm, not sure

w.r.t. the `/etc/hosts` entry; that's expected as of docker 1.10.0, an embedded DNS server is used for discovery / container resolution instead of `/etc/hosts` https://github.com/docker/docker/pull/19198, so you won't see entries being added in `/etc/hosts` for user-defined networks (as are used by compose V2 files)

**Comment by CGavrila** [comment]
So I'm not exactly sure if this is related since it would look like it at first glance, but:

**Comment by CGavrila** [comment]
Not sure what I did, but the problem went away.

This definitely was not some sort of intermittent problem or anything like that, I simply played around in the networking config until it started working, but I cannot remember what changes I did.

**Comment by thaJeztah** [comment]
Good to hear, thanks for reporting back @CGavrila!
